### PR TITLE
fix: allow expired leases to be overwritten

### LINKED ISSUE #826: Lease is stuck even though no parallel jobs are in use when running network deploy
## Description

branch: latest and greatest use-nodejs-cached-version
It started to hang at Acquire lease step:
clue: Acquire lease - lease exists, attempting again in 20000 seconds

### Logs
```
node /home/hashsphere1/.nvm/versions/node/v22.9.0/bin/solo network deploy -i node0,node1,node2,node3,node4,node5,node6 --solo-chart-version v0.34.0 --values-file /home/hashsphere1/workspaces/10nodes/solo/examples/custom-network-config/init-containers-values.yaml --settings-txt /home/hashsphere1/workspaces/10nodes/solo/examples/custom-network-config/settings.txt --log4j2-xml ./log4j2.xml -n solo-hashsphere1

******************************* Solo *********************************************
Version            : 0.99.0
Kubernetes Context    : gke_hashsphere-staging_us-central1_sphere-load-test-us-central
Kubernetes Cluster    : gke_hashsphere-staging_us-central1_sphere-load-test-us-central
Kubernetes Namespace    : solo-hashsphere1
**********************************************************************************
(node:2969811) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
❯ Initialize
 ⠹ Acquire lease - lease exists, attempting again in 20000 seconds, attempt: 4/10
◼ Copy gRPC TLS Certificates
◼ Check if cluster setup chart is installed
◼ Prepare staging directory
◼ Copy node keys to secrets
◼ Install chart 'solo-deployment'
◼ Check node pods are running
◼ Check proxy pods are running
◼ Check auxiliary pods are ready
```

### PR DESCRIPTION
**Pull request by nathanklick**
## Description

This pull request changes the following:

* Fixes an issue where expired leases could permanently prevent the acquisition of another lease.

### Related Issues

* Closes #826

### CODE REVIEW

**Review by alex-kuzmin-hg — APPROVED** [review]
Love it

**Comment by alex-kuzmin-hg** [comment]
Confirmed the fix in this branch:

```
task -t Taskfile.yml clean
task: [solo:node:stop] npm run solo-test -- node stop --namespace "${SOLO_NAMESPACE}" --node-aliases-unparsed node0,node1,node2,node3,node4,node5,node6 
[solo:node:stop] 
[solo:node:stop] > @hashgraph/solo@0.99.0 solo-test
[solo:node:stop] > node --no-deprecation --no-warnings --loader ts-node/esm solo.ts node stop --namespace solo-hashsphere1 --node-aliases-unparsed node0,node1,node2,node3,node4,node5,node6
[solo:node:stop] 
[solo:node:stop] 
[solo:node:stop] ******************************* Solo *********************************************
[solo:node:stop] Version			: 0.99.0
[solo:node:stop] Kubernetes Context	: gke_hashsphere-staging_us-central1_sphere-load-test-us-central
[solo:node:stop] Kubernetes Cluster	: gke_hashsphere-staging_us-central1_sphere-load-test-us-central
[solo:node:stop] Kubernetes Namespace	: solo-hashsphere1
[solo:node:stop] **********************************************************************************
[solo:node:stop] ❯ Initialize
[solo:node:stop] ❯ Acquire lease
[solo:node:stop] ✔ Acquire lease - lease acquired successfully, attempt: 1/10
[solo:node:stop] ✔ Initialize
[solo:node:stop] ❯ Identify network pods
[solo:node:stop] ❯ Check network pod: node0
```

File: src/core/lease_manager.ts [hunk 1]
```diff
@@ -137,8 +141,11 @@ export class LeaseManager {
         `, attempt: ${chalk.cyan(attempt.toString())}/${chalk.cyan(maxAttempts.toString())}`
 
       await sleep(LEASE_ACQUIRE_RETRY_TIMEOUT)
+      lease = await this.tryAcquireLease(username, leaseName, namespace, task, title)
+    }
 
-      return this.acquireLeaseOrRetry(username, leaseName, namespace, task, title, attempt)
+    if (lease) {
+      await this.k8.deleteNamespacedLease(leaseName, namespace)
```
Comment (leninmehedy) [inline src/core/lease_manager.ts:148]: it appears we are deleting the lease that has just been acquired earlier. I believe we need to check if the lease is expired or not before attempting to delete it.

```suggestion
    if (!!lease && this.isLeaseExpired(lease)) {
      await this.k8.deleteNamespacedLease(leaseName, namespace)
    }
```

File: src/core/lease_manager.ts [hunk 2]
```diff
@@ -154,4 +161,18 @@ export class LeaseManager {
     if (!await this.k8.hasNamespace(namespace)) return null
     return namespace
   }
+
+  private isLeaseExpired (lease: V1Lease) : boolean {
+    const now = Date.now()
+    const duration = lease.spec?.leaseDurationSeconds ? lease.spec?.leaseDurationSeconds : 20
```
Comment (leninmehedy) [inline src/core/lease_manager.ts:167]: Would be good to have 20 sec as a constant

**Review by leninmehedy — APPROVED** [review]
Probably worth reviewing the code, refactor or add comments for clarity.

File: src/core/lease_manager.ts (hunk 1 above)
Comment (leninmehedy) [reply src/core/lease_manager.ts:148]: ok, understood it now. If we are here, it must be expired, so we delete it.

File: src/core/lease_manager.ts (hunk 2 above)
Comment (leninmehedy) [reply src/core/lease_manager.ts:167]: nitpik, no need to wait for another CI run. We can fix it later.

**Pull request merged**